aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -1,516 +1,7 @@
-    return;
-  }
-
-  accscalar_t gamma = weight.size(0) > 0 ? static_cast<accscalar_t>(weight[plane]) : static_cast<accscalar_t>(1);
-  accscalar_t beta = bias.size(0) > 0 ? static_cast<accscalar_t>(bias[plane]) : static_cast<accscalar_t>(0);
-  accscalar_t mean = static_cast<accscalar_t>(mean_[plane]);
-  accscalar_t invstd;
-  if (train) {
-    invstd = var_or_invstd[plane];
-  } else {
-    invstd = static_cast<accscalar_t>(1) / device_sqrt(static_cast<accscalar_t>(var_or_invstd[plane]) + epsilon);
-  }
-
-  index_t bs = input.size(0);
-  index_t fs = input.size(2);
-
-  index_t bstep  = blockDim.y * gridDim.y;
-  for (index_t batch = threadIdx.y + blockIdx.y * blockDim.y; batch < bs; batch += bstep) {
-    auto o = output[batch][plane];
-    auto i = input[batch][plane];
-    for (index_t feature = threadIdx.x; feature < fs; feature += blockDim.x) {
-      o[feature] = static_cast<scalar_t>(gamma * (i[feature] - mean) * invstd + beta);
-    }
-  }
-}
-
-
-template <typename scalar_t, typename accscalar_t, typename index_t>
-__global__ void batch_norm_collect_statistics_kernel(
-    const PackedTensorAccessor<scalar_t, 3, RestrictPtrTraits, index_t> input,
-    const accscalar_t epsilon,
-    const accscalar_t momentum,
-    PackedTensorAccessor<scalar_t, 1, RestrictPtrTraits, index_t> running_mean,
-    PackedTensorAccessor<scalar_t, 1, RestrictPtrTraits, index_t> running_var,
-    PackedTensorAccessor<accscalar_t, 1, RestrictPtrTraits, index_t> save_mean,
-    PackedTensorAccessor<accscalar_t, 1, RestrictPtrTraits, index_t> save_invstd) {
-
-  __shared__ int shared_n[2 * 2 * WARP_SIZE + WARP_SIZE];
-
-  int plane = blockIdx.x;
-  int N = input.size(0) * input.size(2);
-  int tid = threadIdx.x + threadIdx.y * blockDim.x;
-
-  // Compute the mean and variance across (batch, x/y/z)
-  // this uses the Welford (in the for loop)/parallel algorithm (to sum across the block)
-  // https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_Online_algorithm
-  // and the parallel algorithm on the same page.
-  // We use two shuffles to reduce across the entire block.
-  // https://devblogs.nvidia.com/faster-parallel-reductions-kepler/ has a description.
-  accscalar_t* shared_avg_var = (accscalar_t*) &shared_n[WARP_SIZE];
-
-  // first the reductions each thread does separately
-  accscalar_t avg = 0;
-  accscalar_t var_n = 0;
-  int n = 0;
-  for (int batch = threadIdx.y; batch < input.size(0); batch += blockDim.y) {
-    for (int x = threadIdx.x; x < input.size(2); x += blockDim.x) {
-      accscalar_t v = input[batch][plane][x];
-      accscalar_t d1 = v - avg;
-      n++;
-      avg += d1 / n;
-      var_n += d1 * (v - avg);
-    }
-  }
-
-  // first warpSum to get one value per thread to
-  // one value per warp
-  for (int i = 0; i < getMSB(WARP_SIZE); ++i) {
-    accscalar_t o_avg = WARP_SHFL_XOR(avg, 1 << i, WARP_SIZE);
-    int o_n = WARP_SHFL_XOR(n, 1 << i, WARP_SIZE);
-    if (n + o_n > 0) {
```

Comment:
This is the real code change, not all threads in warp calls WARP_SHFL_XOR causes cuda hange on volta.